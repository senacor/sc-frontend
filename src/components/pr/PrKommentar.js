import React from 'react';
import { withStyles } from 'material-ui/styles/index';
import List, { ListItem, ListItemText } from 'material-ui/List';
import TextField from 'material-ui/TextField';
import { connect } from 'react-redux';
import Collapse from 'material-ui/transitions/Collapse';
import MenuItem from 'material-ui/Menu/MenuItem';
import Select from 'material-ui/Select';
import FormControl from 'material-ui/Form/FormControl';
import * as actions from '../../actions';

const styles = theme => ({
  nestedText: {
    paddingLeft: '30px',
    paddingRight: '30px'
  },
  bootstrapInput: {
    borderRadius: 4,
    backgroundColor: theme.palette.common.white,
    border: '1px solid #ced4da',
    fontSize: 16,
    padding: '10px 12px',
    '&:focus': {
      borderColor: '#80bdff',
      boxShadow: '0 0 0 0.2rem rgba(0,123,255,.25)'
    }
  },
  containerListItem: {
    display: 'flex'
  },
  nestedListItem: {
    paddingLeft: '30px',
    width: '80%'
  },

  nestedNumber: {
    width: '20%'
  }
});

class PrKommentar extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      rating: this.props.prById.prRatingSet.find(
        rating => rating.prRatingDescription === this.props.category
      ).rating,
      comment: this.props.prById.prRatingSet.find(
        rating => rating.prRatingDescription === this.props.category
      ).comment,
      prById: this.props.prById,
      category: this.props.category,
      PROBLEM_ANALYSIS: false,
      WORK_RESULTS: false,
      WORKING_MANNER: false
    };
  }

  handleChange = (prById, category) => event => {
    this.setState({ [event.target.name]: event.target.value });
    let ratings = prById.prRatingSet.find(
      rating => rating.prRatingDescription === category
    );
    event.target.name === 'rating'
      ? this.props.addRating(
          prById,
          category,
          this.state.comment,
          event.target.value,
          ratings.id
        )
      : this.props.addRating(
          prById,
          category,
          event.target.value,
          this.state.rating,
          ratings.id
        );
  };

  translateRatingDescription = ratingDescription => {
    switch (ratingDescription) {
      case 'PROBLEM_ANALYSIS':
        return 'Problemanalyse und -lösung';
      case 'WORK_RESULTS':
        return 'Arbeitsergebnisse';
      case 'WORKING_MANNER':
        return 'Arbeitsweise';
      default:
        return 'Arbeitsweise';
    }
  };

  handleClick = value => {
    switch (value) {
      case 'WORKING_MANNER':
        this.setState({
          WORKING_MANNER: !this.state.WORKING_MANNER,
          PROBLEM_ANALYSIS: false,
          WORK_RESULTS: false
        });
        break;
      case 'WORK_RESULTS':
        this.setState({
          WORK_RESULTS: !this.state.WORK_RESULTS,
          PROBLEM_ANALYSIS: false,
          WORKING_MANNER: false
        });
        break;
      case 'PROBLEM_ANALYSIS':
        this.setState({
          PROBLEM_ANALYSIS: !this.state.PROBLEM_ANALYSIS,
          WORK_RESULTS: false,
          WORKING_MANNER: false
        });
        break;
      default:
        return;
    }
  };

  categoryText = category => {
    switch (category) {
      case 'PROBLEM_ANALYSIS':
        return {
          categoryDescription:
            'Problemverständnis und Analysestruktur | Relevanz und Korrektheit Analyseergebnisse | korrekte Ableitung von Implikationen',
          positionDescription:
            'Scoping Problem- und Analysekontext; Entwicklung Analysevorgehen, Durchführung von Analysen inkl. Anleitung von Teammitgliedern; Ableitung Implikationen'
        };
      case 'WORKING_MANNER':
        return {
          categoryDescription:
            'Pro-aktive Verantwortungsübernahme | pragmatische Herangehensweise | Qualitätsanspruch (Drive for Excellence)',
          positionDescription:
            'Übernahme von Verantwortung für eigene Endprodukte; Kalibrierung von Vorgehen und Aufwand gegen Lieferzusagen und Machbarkeit; Absicherung der Qualität der verantworteten Endprodukte'
        };

      case 'WORK_RESULTS':
        return {
          categoryDescription:
            'Geeignete und fehlerfreie Ergebnisse | Einhaltung eigener Lieferzusagen | innovative Lösungen',
          positionDescription:
            'Scoping, Strukturierung und termingerechte Erstellung komplexerer Endprodukte in hoher Qualität inkl. Integration von Zulieferungen Dritter; Vorstellung Ergebnisse bei Kunden-Mitarbeitern/Projektleitern'
        };

      default:
        return;
    }
  };

  render() {
    const { prById, category, classes } = this.props;
    return (
      <div>
        <div className={classes.containerListItem}>
          <ListItem
            button
            className={classes.nestedListItem}
            onClick={() => this.handleClick(category)}
          >
            <ListItemText
              secondary={this.translateRatingDescription(category)}
            />
          </ListItem>
          <ListItem className={classes.nestedNumber}>
            <FormControl>
              <Select
                value={this.state.rating}
                onChange={this.handleChange(prById, category)}
                displayEmpty
                name="rating"
              >
                <MenuItem value="">
                  <em>None</em>
                </MenuItem>
                <MenuItem value={1}>1</MenuItem>
                <MenuItem value={2}>2</MenuItem>
                <MenuItem value={3}>3</MenuItem>
                <MenuItem value={4}>4</MenuItem>
                <MenuItem value={5}>5</MenuItem>
              </Select>
            </FormControl>
          </ListItem>
        </div>
        <Collapse in={this.state[category]} timeout="auto" unmountOnExit>
          <List component="div" disablePadding className={classes.nestedText}>
            <ListItem>
              <TextField
                id="multiline-flexible"
                label="comment"
                multiline
                fullWidth
                rowsMax="4"
                value={this.state.comment ? this.state.comment : ''}
                onChange={this.handleChange(prById, category)}
                InputProps={{
                  disableUnderline: true,
                  name: 'comment',
                  classes: {
                    input: classes.bootstrapInput
                  }
                }}
                InputLabelProps={{
                  shrink: true
                }}
              />
            </ListItem>
            <ListItem />
            <ListItem>
              <ListItemText
                secondary={this.categoryText(category).categoryDescription}
              />
            </ListItem>
            <ListItem>
              <ListItemText
                secondary={this.categoryText(category).positionDescription}
              />
            </ListItem>
          </List>
        </Collapse>
      </div>
    );
  }
}

export const StyledComponent = withStyles(styles)(PrKommentar);
export default connect(
  state => ({
    prRating: state.prRatings.prRating
  }),
  {
    addRating: actions.addRating
  }
)(StyledComponent);
