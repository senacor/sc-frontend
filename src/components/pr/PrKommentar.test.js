import React from 'react';
import { StyledComponent } from './PrKommentar';
import { createShallow } from '@material-ui/core/test-utils';

describe('PrKommentar Component', () => {
  let shallow = createShallow({ dive: true });

  const prById = {
    id: 1,
    employee: {
      id: 1,
      login: 'lschäfer',
      firstName: 'Lionel',
      lastName: 'Schäfer',
      title: 'DR',
      email: 'lionel.schäfer@senacor.com',
      entryDate: '2004-05-10',
      salaries: [
        {
          workingHoursPerWeek: 40,
          ote: 45000,
          fte: 1,
          basicSalary: 40500,
          validFrom: '2004-05-10'
        },
        {
          workingHoursPerWeek: 40,
          ote: 50000,
          fte: 1,
          basicSalary: 40000,
          validFrom: '2005-05-10'
        }
      ],
      employment: {
        endOfProbationPeriod: '2004-05-10',
        jobGradings: [
          {
            positionId: 1,
            salaryLevel: 1,
            validFrom: '2004-05-09',
            competence: 'DEVELOPMENT'
          },
          {
            positionId: 2,
            salaryLevel: 2,
            validFrom: '2005-05-09',
            competence: 'DEVELOPMENT'
          }
        ],
        leaves: {
          unpaidLeave: [
            {
              from: '2004-05-10',
              to: '2005-03-31'
            }
          ],
          maternityLeave: [
            {
              from: '2004-05-10',
              to: '2004-10-11'
            }
          ],
          parentalLeave: [
            {
              from: '2004-05-10',
              to: '2004-08-28'
            }
          ],
          sabbatical: [
            {
              from: '2004-05-10',
              to: '2004-06-07'
            },
            {
              from: '2004-05-10',
              to: '2005-02-10'
            },
            {
              from: '2004-05-10',
              to: '2005-03-26'
            }
          ]
        }
      }
    },
    supervisor: 'ttran',
    occasion: 'ON_DEMAND',
    status: 'PREPARATION',
    deadline: '2018-03-14',
    prRatingSet: [
      {
        id: 2,
        prRatingDescription: 'CUSTOMER_RETENTION',
        prRatingCategory: 'IMPACT_ON_CUSTOMER',
        rating: null,
        comment: null
      },
      {
        id: 8,
        prRatingDescription: 'FREE_TEXT_FIELD',
        prRatingCategory: 'OVERALL_ASSESSMENT',
        rating: null,
        comment: null
      },
      {
        id: 3,
        prRatingDescription: 'WORKING_MANNER',
        prRatingCategory: 'PERFORMANCE_IN_PROJECT',
        rating: 1,
        comment: 'nestojjkfgjdkghjrighjfkr'
      },
      {
        id: 1,
        prRatingDescription: 'WORK_RESULTS',
        prRatingCategory: 'PERFORMANCE_IN_PROJECT',
        rating: 2,
        comment: 'kköhhjcsss'
      },
      {
        id: 4,
        prRatingDescription: 'LEADERSHIP',
        prRatingCategory: 'IMPACT_ON_TEAM',
        rating: null,
        comment: null
      },
      {
        id: 9,
        prRatingDescription: 'TEAMWORK',
        prRatingCategory: 'IMPACT_ON_TEAM',
        rating: null,
        comment: null
      },
      {
        id: 5,
        prRatingDescription: 'PROBLEM_ANALYSIS',
        prRatingCategory: 'PERFORMANCE_IN_PROJECT',
        rating: 5,
        comment: 'kvvvuierkfjrfhkrehghfj'
      },
      {
        id: 6,
        prRatingDescription: 'CONTRIBUTION_TO_COMPANY_DEVELOPMENT',
        prRatingCategory: 'IMPACT_ON_COMPANY',
        rating: null,
        comment: null
      },
      {
        id: 7,
        prRatingDescription: 'CUSTOMER_INTERACTION',
        prRatingCategory: 'IMPACT_ON_CUSTOMER',
        rating: null,
        comment: null
      }
    ],
    taskSet: [
      {
        id: 1,
        title: 'PR MMustermann: Bewerter bestimmen'
      },
      {
        id: 2,
        title: 'PR MMustermann: Terminierung'
      }
    ],
    _links: {
      self: {
        href: 'http://localhost:8010/api/v1/prs/1'
      }
    }
  };

  it('displays the PrKommentar', () => {
    const component = shallow(
      <StyledComponent prById={prById} category="PROBLEM_ANALYSIS" />
    );

    expect(component).toMatchSnapshot();
  });
});
