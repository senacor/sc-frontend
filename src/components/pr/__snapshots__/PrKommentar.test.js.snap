// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrKommentar Component should display Customer interaction 1`] = `
<div>
  <div
    className="PrKommentar-containerListItem-3"
  >
    <WithStyles(ListItem)
      button={true}
      className="PrKommentar-nestedListItem-4"
      onClick={[Function]}
    >
      <WithStyles(ListItemText)
        className={Object {}}
        secondary="Kundeninteraktion und -veränderung"
      />
      <WithStyles(Icon)
        className="PrKommentar-icon-6"
      >
        comment
      </WithStyles(Icon)>
    </WithStyles(ListItem)>
    <WithStyles(ListItem)
      className="PrKommentar-nestedNumber-7"
    >
      <WithStyles(FormControl)
        className="PrKommentar-formControlFTF-10"
      >
        <WithStyles(Select)
          displayEmpty={true}
          name="rating"
          onChange={[Function]}
          value={4}
        >
          <WithStyles(MenuItem)
            value={1}
          >
            1
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value={2}
          >
            2
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value={3}
          >
            3
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value={4}
          >
            4
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value={5}
          >
            5
          </WithStyles(MenuItem)>
        </WithStyles(Select)>
      </WithStyles(FormControl)>
    </WithStyles(ListItem)>
  </div>
  <WithStyles(Collapse)
    in={false}
    timeout="auto"
    unmountOnExit={true}
  >
    <WithStyles(List)
      className="PrKommentar-nestedText-1"
      component="div"
      disablePadding={true}
    >
      <WithStyles(ListItem)>
        <TextField
          InputLabelProps={
            Object {
              "shrink": true,
            }
          }
          InputProps={
            Object {
              "classes": Object {
                "input": "PrKommentar-bootstrapInput-2",
              },
              "disableUnderline": true,
              "name": "comment",
            }
          }
          fullWidth={true}
          id="CUSTOMER_INTERACTION"
          label="Kommentar"
          multiline={true}
          onChange={[Function]}
          required={false}
          rowsMax="4"
          select={false}
          value="iii"
        />
      </WithStyles(ListItem)>
      <WithStyles(ListItem) />
      <WithStyles(PrSwipePositionDescription)
        category="CUSTOMER_INTERACTION"
      />
    </WithStyles(List)>
  </WithStyles(Collapse)>
</div>
`;

exports[`PrKommentar Component should display ProblemAnalysis 1`] = `
<div>
  <div
    className="PrKommentar-containerListItem-3"
  >
    <WithStyles(ListItem)
      button={true}
      className="PrKommentar-nestedListItem-4"
      onClick={[Function]}
    >
      <WithStyles(ListItemText)
        className={Object {}}
        secondary="Problemanalyse und -lösung"
      />
      <WithStyles(Icon)
        className="PrKommentar-icon-6"
      >
        comment
      </WithStyles(Icon)>
    </WithStyles(ListItem)>
    <WithStyles(ListItem)
      className="PrKommentar-nestedNumber-7"
    >
      <WithStyles(FormControl)
        className="PrKommentar-formControlFTF-10"
      >
        <WithStyles(Select)
          displayEmpty={true}
          name="rating"
          onChange={[Function]}
          value={2}
        >
          <WithStyles(MenuItem)
            value={1}
          >
            1
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value={2}
          >
            2
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value={3}
          >
            3
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value={4}
          >
            4
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value={5}
          >
            5
          </WithStyles(MenuItem)>
        </WithStyles(Select)>
      </WithStyles(FormControl)>
    </WithStyles(ListItem)>
  </div>
  <WithStyles(Collapse)
    in={false}
    timeout="auto"
    unmountOnExit={true}
  >
    <WithStyles(List)
      className="PrKommentar-nestedText-1"
      component="div"
      disablePadding={true}
    >
      <WithStyles(ListItem)>
        <TextField
          InputLabelProps={
            Object {
              "shrink": true,
            }
          }
          InputProps={
            Object {
              "classes": Object {
                "input": "PrKommentar-bootstrapInput-2",
              },
              "disableUnderline": true,
              "name": "comment",
            }
          }
          fullWidth={true}
          id="PROBLEM_ANALYSIS"
          label="Kommentar"
          multiline={true}
          onChange={[Function]}
          required={false}
          rowsMax="4"
          select={false}
          value="ggg"
        />
      </WithStyles(ListItem)>
      <WithStyles(ListItem) />
      <WithStyles(PrSwipePositionDescription)
        category="PROBLEM_ANALYSIS"
      />
    </WithStyles(List)>
  </WithStyles(Collapse)>
</div>
`;

exports[`PrKommentar Component should display Teamwork 1`] = `
<div>
  <div
    className="PrKommentar-containerListItem-3"
  >
    <WithStyles(ListItem)
      button={true}
      className="PrKommentar-nestedListItem-4"
      onClick={[Function]}
    >
      <WithStyles(ListItemText)
        className={Object {}}
        secondary="Effektives Arbeiten im Team und Teamführung"
      />
      <WithStyles(Icon)
        className="PrKommentar-icon-6"
      >
        comment
      </WithStyles(Icon)>
    </WithStyles(ListItem)>
    <WithStyles(ListItem)
      className="PrKommentar-nestedNumber-7"
    >
      <WithStyles(FormControl)
        className="PrKommentar-formControlFTF-10"
      >
        <WithStyles(Select)
          displayEmpty={true}
          name="rating"
          onChange={[Function]}
          value={1}
        >
          <WithStyles(MenuItem)
            value={1}
          >
            1
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value={2}
          >
            2
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value={3}
          >
            3
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value={4}
          >
            4
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value={5}
          >
            5
          </WithStyles(MenuItem)>
        </WithStyles(Select)>
      </WithStyles(FormControl)>
    </WithStyles(ListItem)>
  </div>
  <WithStyles(Collapse)
    in={false}
    timeout="auto"
    unmountOnExit={true}
  >
    <WithStyles(List)
      className="PrKommentar-nestedText-1"
      component="div"
      disablePadding={true}
    >
      <WithStyles(ListItem)>
        <TextField
          InputLabelProps={
            Object {
              "shrink": true,
            }
          }
          InputProps={
            Object {
              "classes": Object {
                "input": "PrKommentar-bootstrapInput-2",
              },
              "disableUnderline": true,
              "name": "comment",
            }
          }
          fullWidth={true}
          id="TEAMWORK"
          label="Kommentar"
          multiline={true}
          onChange={[Function]}
          required={false}
          rowsMax="4"
          select={false}
          value="fff"
        />
      </WithStyles(ListItem)>
      <WithStyles(ListItem) />
      <WithStyles(PrSwipePositionDescription)
        category="TEAMWORK"
      />
    </WithStyles(List)>
  </WithStyles(Collapse)>
</div>
`;

exports[`PrKommentar Component should display the ContributionToCompanyDevelopment 1`] = `
<div>
  <div
    className="PrKommentar-containerListItem-3"
  >
    <WithStyles(ListItem)
      button={true}
      className="PrKommentar-nestedListItem-4"
      onClick={[Function]}
    >
      <WithStyles(ListItemText)
        className={Object {}}
        secondary="Beitrag zur Unternehmensentwicklung"
      />
      <WithStyles(Icon)
        className="PrKommentar-icon-6"
      >
        comment
      </WithStyles(Icon)>
    </WithStyles(ListItem)>
    <WithStyles(ListItem)
      className="PrKommentar-nestedNumber-7"
    >
      <WithStyles(FormControl)
        className="PrKommentar-formControlFTF-10"
      >
        <WithStyles(Select)
          displayEmpty={true}
          name="rating"
          onChange={[Function]}
          value={3}
        >
          <WithStyles(MenuItem)
            value={1}
          >
            1
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value={2}
          >
            2
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value={3}
          >
            3
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value={4}
          >
            4
          </WithStyles(MenuItem)>
          <WithStyles(MenuItem)
            value={5}
          >
            5
          </WithStyles(MenuItem)>
        </WithStyles(Select)>
      </WithStyles(FormControl)>
    </WithStyles(ListItem)>
  </div>
  <WithStyles(Collapse)
    in={false}
    timeout="auto"
    unmountOnExit={true}
  >
    <WithStyles(List)
      className="PrKommentar-nestedText-1"
      component="div"
      disablePadding={true}
    >
      <WithStyles(ListItem)>
        <TextField
          InputLabelProps={
            Object {
              "shrink": true,
            }
          }
          InputProps={
            Object {
              "classes": Object {
                "input": "PrKommentar-bootstrapInput-2",
              },
              "disableUnderline": true,
              "name": "comment",
            }
          }
          fullWidth={true}
          id="CONTRIBUTION_TO_COMPANY_DEVELOPMENT"
          label="Kommentar"
          multiline={true}
          onChange={[Function]}
          required={false}
          rowsMax="4"
          select={false}
          value="hhh"
        />
      </WithStyles(ListItem)>
      <WithStyles(ListItem) />
      <WithStyles(PrSwipePositionDescription)
        category="CONTRIBUTION_TO_COMPANY_DEVELOPMENT"
      />
    </WithStyles(List)>
  </WithStyles(Collapse)>
</div>
`;
